### PR TITLE
feat(android): presentPermissionsPickerAsync in media-library

### LINKED ISSUE #29537: [expo-media-library][Android] Add support for presentPermissionsPickerAsync for Android
### Minimal reproducible example

https://github.com/RRaideRR/expo-media-library-selective-permissions

### What platform(s) does this occur on?

Android

### Did you reproduce this issue in a development build?

Yes

### Summary

Android introduced support for partial photo access a couple of months ago (https://developer.android.com/about/versions/14/changes/partial-photo-video-access). If the user gave `accessPriviliges=limited`, expo-media-library provides no way to ask him again to adjust his selection. `expo-image-picker` resolved that issue with #28510. 

My current solution is to re-ask the user to give his permission on Android. However, this does not trigger the `MediaLibrary.addListener`. So there is no chance to recognise asset changes on Android.

```
  if (Platform.OS === 'ios') {
    MediaLibrary.presentPermissionsPickerAsync();
    return;
  }

  if (Platform.OS === 'android') {
    requestPermission();
  }
```

In my naive imagination it would be nice to have the `presentPermissionsPickerAsync` method work for Android as well (in a similar manner compared to iOS).

iOS:
https://github.com/expo/expo/assets/3411292/5d2379ca-1734-4574-9925-f0a5939d7bd9

Android:

https://github.com/expo/expo/assets/3411292/bf6c5136-af26-4dfa-a54c-cf0be4781e25

### Environment

  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.5
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.20.2 - ~/.nvm/versions/node/v18.20.2/bin/node
      Yarn: 1.22.19 - ~/.yarn/bin/yarn
      npm: 10.5.0 - ~/.nvm/versions/node/v18.20.2/bin/npm
      Watchman: 2024.04.29.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/opt/ruby/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.5, iOS 17.5, macOS 14.5, tvOS 17.5, visionOS 1.2, watchOS 10.5
    IDEs:
      Android Studio: 2023.3 AI-233.14808.21.2331.11709847
      Xcode: 15.4/15F31d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~51.0.11 => 51.0.11 
      react: 18.2.0 => 18.2.0 
      react-native: 0.74.2 => 0.74.2 
    npmGlobalPackages:
      eas-cli: 9.1.0
    Expo Workflow: managed

### Expo Doctor Diagnostics

✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check native tooling versions
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check for issues with metro config
✔ Check npm/ yarn versions
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check that packages match versions required by installed Expo SDK
✔ Check that native modules do not use incompatible support packages
✔ Check for legacy global CLI installed locally
✔ Check that native modules use compatible support package versions for installed Expo SDK

Didn't find any issues with the project!

### PR DESCRIPTION
**Pull request by vonovak**
# Why

closes #29537

# How

- `presentPermissionsPickerAsync` is altered on android 14+ to call `requestPermissionsAsync` 
- docs are updated

# Test Plan

I tested this manually. On Android < 14, this api is not available. On newer versions, these screenshots apply:

<details>
  <summary>Click me</summary>

Calling `presentPermissionsPickerAsync ` without params.

initial

Calling `presentPermissionsPickerAsync ` with `['video']` parameter 

presentPermissionsPickerAsync-video

</details>

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (no updates needed).

### CODE REVIEW

File: packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MediaLibraryModule.kt [hunk 1]
```diff
@@ -361,28 +361,32 @@ class MediaLibraryModule : Module() {
...
       WRITE_EXTERNAL_STORAGE.takeIf { shouldAddWriteExternalStorage },
       READ_EXTERNAL_STORAGE.takeIf { !writeOnly && !shouldAddGranularPermissions },
       ACCESS_MEDIA_LOCATION.takeIf { shouldAddMediaLocationAccess },
-      *getGranularPermissions(writeOnly, shouldAddGranularPermissions, granularPermissions)
+      *getGranularPermissions(shouldIncludeGranular, granularPermissions)
     ).toTypedArray()
   }
 
   @SuppressLint("InlinedApi")
-  private fun getGranularPermissions(writeOnly: Boolean, shouldAdd: Boolean, granularPermissions: List<GranularPermission>): Array<String> {
```
Comment (vonovak) [inline packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MediaLibraryModule.kt:373]: just a tiny change: 2 boolean params reduced to 1

File: packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MediaLibraryModule.kt [hunk 2]
```diff
@@ -490,7 +494,7 @@ class MediaLibraryModule : Module() {
     private fun getAssetsTotalCount(mediaType: Int): Int =
       context.contentResolver.query(
         EXTERNAL_CONTENT_URI,
-        null,
+        arrayOf(),
```
Comment (vonovak) [inline packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MediaLibraryModule.kt:497]: https://developer.android.com/reference/android/content/ContentResolver#query(android.net.Uri,%20java.lang.String[],%20java.lang.String,%20java.lang.String[],%20java.lang.String,%20android.os.CancellationSignal)

> String: A list of which columns to return. Passing null will return all columns, which is inefficient.


here we don't care about any of the columns, we just care about the number of items in the cursor

File: packages/expo-media-library/ios/MediaLibraryModule.swift [hunk 3]
```diff
@@ -65,7 +65,7 @@ public class MediaLibraryModule: Module, PhotoLibraryObserverHandler {
     }
 
     AsyncFunction("presentPermissionsPickerAsync") {
-      if #available(iOS 15.0, *) {
+      if #available(iOS 14.0, *) {
```
Comment (vonovak) [inline packages/expo-media-library/ios/MediaLibraryModule.swift:68]: the docs say this correctly

File: packages/expo-media-library/src/MediaLibrary.ts [hunk 4]
```diff
@@ -474,16 +483,21 @@ export const usePermissions = createPermissionHook<
 
 // @needsAudit
 /**
- * __Available only on iOS >= 14.__ Allows the user to update the assets that your app has access to.
+ * __Available only on iOS >= 14 and Android >= 14.__ Allows the user to update the assets that your app has access to.
```
Comment (Simek) [inline packages/expo-media-library/src/MediaLibrary.ts:486]:
```suggestion
 * Allows the user to update the assets that your app has access to.
```

We can use the platform tags to indicate the required platform versions:
```
@platform android 14+
@platform ios 14+
```


In long term, we would like to show all supported platforms for every API entry. The process will be automated, based on platforms set for the page, but in cases like that we would need to add those params to overwrite the default value for the package anyway.

A bit more information about params we use in source comments:
* https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#general-approach

File: packages/expo-media-library/src/MediaLibrary.ts [hunk 5]
```diff
@@ -16,19 +16,28 @@ export type PermissionResponse = EXPermissionResponse & {
...
+ * @platform android 13+
  */
 export type GranularPermission = 'audio' | 'photo' | 'video';
 
 export type MediaTypeValue = 'audio' | 'photo' | 'video' | 'unknown';
+
+/**
+ * On Android 14+, when calling [`presentPermissionsPickerAsync()`](#medialibrarypresentpermissionspickerasyncmediatypes),
+ * use this to specify the type(s) of media that the user will be granting access to.
+ * By default, this presents a list that shows both photos and videos.
```
Comment (Simek) [inline packages/expo-media-library/src/MediaLibrary.ts:36]:
```suggestion
```

This should be a part of `presentPermissionsPickerAsync` param description.

File: packages/expo-media-library/src/MediaLibrary.ts [hunk 6]
```diff
@@ -16,19 +16,28 @@ export type PermissionResponse = EXPermissionResponse & {
...
- * @platform android API 33+
+ * @platform android 13+
  */
 export type GranularPermission = 'audio' | 'photo' | 'video';
 
 export type MediaTypeValue = 'audio' | 'photo' | 'video' | 'unknown';
+
+/**
+ * On Android 14+, when calling [`presentPermissionsPickerAsync()`](#medialibrarypresentpermissionspickerasyncmediatypes),
+ * use this to specify the type(s) of media that the user will be granting access to.
```
Comment (Simek) [inline packages/expo-media-library/src/MediaLibrary.ts:35]:
```suggestion
 * Represents the possible types of media that app will ask the OS to get access to when calling [`presentPermissionsPickerAsync()`](#medialibrarypresentpermissionspickerasyncmediatypes).
```

Let's avoid the platform repetition, focus on what the types represents, and move some parts out.

File: packages/expo-media-library/src/MediaLibrary.ts [hunk 7]
```diff
@@ -474,16 +483,21 @@ export const usePermissions = createPermissionHook<
 
 // @needsAudit
 /**
- * __Available only on iOS >= 14.__ Allows the user to update the assets that your app has access to.
+ * __Available only on iOS >= 14 and Android >= 14.__ Allows the user to update the assets that your app has access to.
  * The system modal is only displayed if the user originally allowed only `limited` access to their
  * media library, otherwise this method is a no-op.
- * @return A promise that either rejects if the method is unavailable (meaning the device is not
- * running iOS >= 14), or resolves to `void`.
+ * @return A promise that either rejects if the method is unavailable, or resolves to `void`.
```
Comment (Simek) [inline packages/expo-media-library/src/MediaLibrary.ts:489]:
```suggestion
 * @param mediaTypes [FILL_ME_OUT]
 * @return A promise that either rejects if the method is unavailable, or resolves to `void`.
```

Let's add new param description.

**Review by Simek — COMMENTED** [review]
Few nits for the source docs.

**Review by lukmccall — APPROVED** [review]

**Review by Simek — APPROVED** [review]
Thanks for the quick updates! 👍

**Review by amandeepmittal — APPROVED** [review]
Docs changes look good to me, thanks! 👍

**Comment by vonovak** [comment]
Alan is vacationing, so I merged.

**Pull request merged**